src/beast/beast/asio/append_buffers.h

```diff
@@ -0,0 +1,516 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of Beast: https://github.com/vinniefalco/Beast
+    Copyright 2013, Vinnie Falco <vinnie.falco@gmail.com>
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef BEAST_ASIO_APPEND_BUFFERS_H_INLUDED
+#define BEAST_ASIO_APPEND_BUFFERS_H_INLUDED
+
+#include <boost/asio/buffer.hpp>
+#include <cstdint>
+#include <iterator>
+#include <stdexcept>
+#include <tuple>
+#include <utility>
+
+namespace beast {
+
+namespace detail {
+
+template<class ValueType, class... Bs>
+class append_buffers_helper
+{
+    std::tuple<Bs...> bs_;
+
+public:
+    using value_type = ValueType;
+
+    class const_iterator;
+
+    append_buffers_helper(append_buffers_helper&&) = default;
+    append_buffers_helper(append_buffers_helper const&) = default;
+    append_buffers_helper& operator=(append_buffers_helper&&) = default;
+    append_buffers_helper& operator=(append_buffers_helper const&) = default;
+
+    explicit
+    append_buffers_helper(Bs const&... bs)
+        : bs_(bs...)
+    {
+    }
+
+    const_iterator
+    begin() const;
+
+    const_iterator
+    end() const;
+};
+
+template<class ValueType, class... Bs>
+class append_buffers_helper<
+    ValueType, Bs...>::const_iterator
+{
+    template<class U>
+    static
+    inline
+    std::size_t constexpr
+    max_sizeof()
+    {
+        return sizeof(U);
+    }
+
+    template<class U0, class U1, class... Us>
+    static
+    inline
+    std::size_t constexpr
+    max_sizeof()
+    {
+        return
+            max_sizeof<U0>() > max_sizeof<U1, Us...>() ?
+            max_sizeof<U0>() : max_sizeof<U1, Us...>();
+    }
+
+    std::size_t n_;
+    std::tuple<Bs...> const* bs_;
+    std::array<std::uint8_t,
+        max_sizeof<typename Bs::const_iterator...>()> buf_;
```

Comment:
Style nitpick:  If you aren't going to put your data members first, you might flag them with comments to make them easier to find:
````
//  data members begin
std::size_t n_;
...
//  data members end
````